src/reflect/scala/reflect/internal/tpe/TypeMaps.scala

```diff
@@ -758,21 +775,21 @@ private[internal] trait TypeMaps {
     }
   }
 
-  /** A map to implement the `substSym` method. */
-  class SubstSymMap(from: List[Symbol], to: List[Symbol]) extends SubstMap(from, to) {
-    def this(pairs: (Symbol, Symbol)*) = this(pairs.toList.map(_._1), pairs.toList.map(_._2))
+  abstract class SubstMap[T >: Null](override val from: List[Symbol], to: List[T]) extends AbstractSubstMap[T] {
+    // OPT this check was 2-3% of some profiles, demoted to -Xdev
+    if (isDeveloper) assert(sameLength(from, to), "Unsound substitution from "+ from +" to "+ to)
 
-    protected def toType(fromtp: Type, sym: Symbol) = fromtp match {
-      case TypeRef(pre, _, args) => copyTypeRef(fromtp, pre, sym, args)
-      case SingleType(pre, _) => singleType(pre, sym)
-    }
-    @tailrec private def subst(sym: Symbol, from: List[Symbol], to: List[Symbol]): Symbol = (
-      if (from.isEmpty) sym
+    protected def find(sym: Symbol): T = findAux(sym, from, to)
+
+    @tailrec private def findAux(sym: Symbol, from: List[Symbol], to: List[T]): T =
+      if (from.isEmpty) (null: T)
       // else if (to.isEmpty) error("Unexpected substitution on '%s': from = %s but to == Nil".format(sym, from))
       else if (matches(from.head, sym)) to.head
-      else subst(sym, from.tail, to.tail)
-      )
-    private def substFor(sym: Symbol) = subst(sym, from, to)
+      else findAux(sym, from.tail, to.tail)
+  }
+
+  abstract class AbstractSubstSymMap extends AbstractSubstMap[Symbol] {
+    protected def to: List[Symbol]
 
     override def apply(tp: Type): Type = (
       if (from.isEmpty) tp
```

Comment:
The other [crash](
https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/11075/testReport/junit/test.files/run_/reflection_tags_scala/) is here, but that would also be explained if there's a race with `pop`.